outlines/text/sequences/sequence.py

```diff
@@ -0,0 +1,259 @@
+
+        """
+        num_input_dims = token_ids.ndim
+        probs = self.model(token_ids, attention_mask)
+
+        # Sample `samples` new tokens
+        vocab_size = probs.shape[-1]
+        next_token_ids = vectorized_random_choice(
+            rng, np.arange(vocab_size), probs, samples
+        )
+
+        # Add the missing `num_tokens` and `num_sample` dimensions
+        next_token_ids = np.expand_dims(next_token_ids, -1)
+        token_ids = np.expand_dims(token_ids, 0)
+
+        # Expand the input `token_ids` array to be able to concatenate several
+        # samples.
+        if samples > 1:
+            repetitions = (samples,) + (1,) * num_input_dims
+            token_ids = np.tile(token_ids, repetitions)
+            probs = np.tile(probs, repetitions)
+
+        token_ids = np.concatenate([token_ids, next_token_ids], axis=-1)
+
+        # Merge sample and batch dimensions by removing dimensions of length
+        # 1. The shape of the resulting arrays is `new_batch_shape + (num_tokens,)`
+        # and `new_batch_shape + (vocab_size,)` respectively.
+        token_ids = np.atleast_2d(token_ids.squeeze())
+        probs = np.atleast_2d(probs.squeeze())
+
+        return token_ids, probs
+
+    def expand_attention_mask(
+        self, attention_mask: NDArray[np.int64]
+    ) -> NDArray[np.int64]:
+        """Expand the attention mask after the last completion."""
+        batch_shape = attention_mask.shape[:-1]
+        attention_mask = np.concatenate(
+            [attention_mask, np.broadcast_to([1], batch_shape + (1,))], axis=-1
+        )
+        return attention_mask
+
+    def update_token_ids(
+        self,
+        is_finished: NDArray[np.bool_],
+        token_ids: NDArray[np.int64],
+        token_ids_unfinished: NDArray[np.int64],
+    ) -> NDArray[np.int64]:
+        """Update the array of token ids after the last completion.
+
+        We only generate new tokens for the sequences that are not finished. We thus
+        update the array with the new tokens, and append pad tokens to the finished
+        sequences.
+
+        Parameters
+        ----------
+        is_finished
+            Boolean array that indicates which sequences are finished.
+        token_ids
+            Array that contains the sequences before the generation's last step.
+        token_ids_unfinished
+            Array that contains the sequences of the unfinished sequences
+            after the generation's last step.
+
+        Returns
+        -------
+        An array that contains the updated array that contains the sequences. We append
+        pad tokens to the finished sequences.
+
+        """
+        batch_shape = token_ids.shape[:-1]
+        num_tokens = token_ids.shape[-1]
+        new_token_ids = np.empty(batch_shape + (num_tokens + 1,), dtype=np.int64)
+
+        token_ids_finished = token_ids[is_finished]
+        batch_shape_finished = token_ids_finished.shape[:-1]
+        token_ids_finished = np.concatenate(
```

Comment:
cf comment on Boolean indexing below. Padding could be performed once at the end of generation.